File: crates/dw-config/src/base_dirs.rs
use directories::BaseDirs;
use std::env;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PlatformBaseDirs {
    pub home_dir: PathBuf,
    pub cache_dir: Option<PathBuf>,
    pub config_dir: Option<PathBuf>,
    pub data_dir: Option<PathBuf>,
    pub data_local_dir: Option<PathBuf>,
    pub executable_dir: Option<PathBuf>,
    pub preference_dir: Option<PathBuf>,
    pub runtime_dir: Option<PathBuf>,
    pub state_dir: Option<PathBuf>,
}

impl PlatformBaseDirs {
    pub fn resolve() -> Self {
        if let Some(base_dirs) = BaseDirs::new() {
            return Self {
                home_dir: base_dirs.home_dir().to_path_buf(),
                cache_dir: Some(base_dirs.cache_dir().to_path_buf()),
                config_dir: Some(base_dirs.config_dir().to_path_buf()),
                data_dir: Some(base_dirs.data_dir().to_path_buf()),
                data_local_dir: Some(base_dirs.data_local_dir().to_path_buf()),
                executable_dir: base_dirs.executable_dir().map(PathBuf::from),
                preference_dir: Some(base_dirs.preference_dir().to_path_buf()),
                runtime_dir: base_dirs.runtime_dir().map(PathBuf::from),
                state_dir: base_dirs.state_dir().map(PathBuf::from),
            };
        }

        let home_dir = fallback_home_dir();
        Self {
            cache_dir: fallback_cache_dir(&home_dir),
            config_dir: fallback_config_dir(&home_dir),
            data_dir: fallback_data_dir(&home_dir),
            data_local_dir: fallback_data_local_dir(&home_dir),
            executable_dir: None,
            preference_dir: fallback_preference_dir(&home_dir),
            runtime_dir: None,
            state_dir: fallback_state_dir(&home_dir),
            home_dir,
        }
    }

    pub fn default_root(&self) -> PathBuf {
        self.home_dir.join("dev").join("dw")
    }

    pub fn user_config_directory(&self) -> PathBuf {
        if cfg!(windows) {
            return self
                .data_local_dir
                .clone()
                .or_else(|| self.config_dir.clone())
                .unwrap_or_else(|| self.home_dir.clone())
                .join("DevWorkflow");
        }

        self.config_dir
            .clone()
            .unwrap_or_else(|| self.home_dir.join(".config"))
            .join("DevWorkflow")
    }
}

fn fallback_home_dir() -> PathBuf {
    env_value("HOME")
        .or_else(|| env_value("USERPROFILE"))
        .or_else(|| match (env_value("HOMEDRIVE"), env_value("HOMEPATH")) {
            (Some(drive), Some(path)) => Some(format!("{drive}{path}")),
            _ => None,
        })
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("~"))
}

fn fallback_cache_dir(home_dir: &Path) -> Option<PathBuf> {
    if cfg!(windows) {
        return env_value("LOCALAPPDATA").map(PathBuf::from);
    }
    env_value("XDG_CACHE_HOME")
        .map(PathBuf::from)
        .or_else(|| Some(home_dir.join(".cache")))
}

fn fallback_config_dir(home_dir: &Path) -> Option<PathBuf> {
    if cfg!(windows) {
        return env_value("APPDATA")
            .or_else(|| env_value("LOCALAPPDATA"))
            .map(PathBuf::from);
    }
    env_value("XDG_CONFIG_HOME")
        .map(PathBuf::from)
        .or_else(|| Some(home_dir.join(".config")))
}

fn fallback_data_dir(home_dir: &Path) -> Option<PathBuf> {
    if cfg!(windows) {
        return env_value("APPDATA")
            .or_else(|| env_value("LOCALAPPDATA"))
            .map(PathBuf::from);
    }
    env_value("XDG_DATA_HOME")
        .map(PathBuf::from)
        .or_else(|| Some(home_dir.join(".local").join("share")))
}

fn fallback_data_local_dir(home_dir: &Path) -> Option<PathBuf> {
    if cfg!(windows) {
        return env_value("LOCALAPPDATA").map(PathBuf::from);
    }
    fallback_data_dir(home_dir)
}

fn fallback_preference_dir(home_dir: &Path) -> Option<PathBuf> {
    fallback_config_dir(home_dir)
}

fn fallback_state_dir(home_dir: &Path) -> Option<PathBuf> {
    if cfg!(windows) {
        return None;
    }
    env_value("XDG_STATE_HOME")
        .map(PathBuf::from)
        .or_else(|| Some(home_dir.join(".local").join("state")))
}

fn env_value(key: &str) -> Option<String> {
    env::var(key)
        .ok()
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolved_base_dirs_include_home_and_config() {
        let dirs = PlatformBaseDirs::resolve();

        assert!(!dirs.home_dir.as_os_str().is_empty());
        assert!(!dirs.user_config_directory().as_os_str().is_empty());
    }
}

File: internal/config/base_dirs.go
package config

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// PlatformBaseDirs are the platform user directories used by dw. Empty optional
// fields mean the operating system does not define a corresponding directory.
type PlatformBaseDirs struct {
	HomeDir       string `json:"homeDir"`
	CacheDir      string `json:"cacheDir,omitempty"`
	ConfigDir     string `json:"configDir,omitempty"`
	DataDir       string `json:"dataDir,omitempty"`
	DataLocalDir  string `json:"dataLocalDir,omitempty"`
	ExecutableDir string `json:"executableDir,omitempty"`
	PreferenceDir string `json:"preferenceDir,omitempty"`
	RuntimeDir    string `json:"runtimeDir,omitempty"`
	StateDir      string `json:"stateDir,omitempty"`
}

// ResolvePlatformBaseDirs resolves base directories without touching the file
// system. Environment values containing only whitespace are treated as unset.
func ResolvePlatformBaseDirs() PlatformBaseDirs {
	home := resolveHomeDir()
	if runtime.GOOS == "windows" {
		local := firstEnvironment("LOCALAPPDATA")
		roaming := firstEnvironment("APPDATA", "LOCALAPPDATA")
		return PlatformBaseDirs{
			HomeDir:       home,
			CacheDir:      local,
			ConfigDir:     roaming,
			DataDir:       roaming,
			DataLocalDir:  local,
			PreferenceDir: roaming,
		}
	}

	config := firstEnvironment("XDG_CONFIG_HOME")
	if config == "" {
		config = filepath.Join(home, ".config")
	}
	cache := firstEnvironment("XDG_CACHE_HOME")
	if cache == "" {
		cache = filepath.Join(home, ".cache")
	}
	data := firstEnvironment("XDG_DATA_HOME")
	if data == "" {
		data = filepath.Join(home, ".local", "share")
	}
	state := firstEnvironment("XDG_STATE_HOME")
	if state == "" {
		state = filepath.Join(home, ".local", "state")
	}
	executable := firstEnvironment("XDG_BIN_HOME")
	if executable == "" {
		executable = filepath.Join(filepath.Dir(data), "bin")
	}
	runtimeDir := firstEnvironment("XDG_RUNTIME_DIR")
	return PlatformBaseDirs{
		HomeDir:       home,
		CacheDir:      cache,
		ConfigDir:     config,
		DataDir:       data,
		DataLocalDir:  data,
		ExecutableDir: executable,
		PreferenceDir: config,
		RuntimeDir:    runtimeDir,
		StateDir:      state,
	}
}

func (dirs PlatformBaseDirs) DefaultRoot() string {
	return filepath.Join(dirs.HomeDir, "dev", "dw")
}

func (dirs PlatformBaseDirs) UserConfigDirectory() string {
	if runtime.GOOS == "windows" {
		base := dirs.DataLocalDir
		if base == "" {
			base = dirs.ConfigDir
		}
		if base == "" {
			base = dirs.HomeDir
		}
		return filepath.Join(base, "DevWorkflow")
	}
	base := dirs.ConfigDir
	if base == "" {
		base = filepath.Join(dirs.HomeDir, ".config")
	}
	return filepath.Join(base, "DevWorkflow")
}

func resolveHomeDir() string {
	if runtime.GOOS == "windows" {
		if home := firstEnvironment("USERPROFILE"); home != "" {
			return home
		}
		drive, path := environment("HOMEDRIVE"), environment("HOMEPATH")
		if drive != "" && path != "" {
			return drive + path
		}
		if home := firstEnvironment("HOME"); home != "" {
			return home
		}
		return "~"
	}
	if home := firstEnvironment("HOME", "USERPROFILE"); home != "" {
		return home
	}
	drive, path := environment("HOMEDRIVE"), environment("HOMEPATH")
	if drive != "" && path != "" {
		return drive + path
	}
	return "~"
}

func firstEnvironment(keys ...string) string {
	for _, key := range keys {
		if value := environment(key); value != "" {
			return value
		}
	}
	return ""
}

func environment(key string) string {
	return strings.TrimSpace(os.Getenv(key))
}
